File: lib/data/dto/cashout_dto.dart
class CashOutRequest {
  final int coinAmount;
  final String paymentMethod; // paypal | bank_transfer | crypto
  final String paymentDetails;

  const CashOutRequest({
    required this.coinAmount,
    required this.paymentMethod,
    required this.paymentDetails,
  });

  Map<String, dynamic> toJson() => {
        'coinAmount': coinAmount,
        'paymentMethod': paymentMethod,
        'paymentDetails': paymentDetails,
      };
}

class CashOutResponseDto {
  final bool success;
  final String cashOutId;
  final int coinAmount;
  final double cashValue;
  final int newBalance;
  final String status; // pending | processing | completed | failed

  const CashOutResponseDto({
    required this.success,
    required this.cashOutId,
    required this.coinAmount,
    required this.cashValue,
    required this.newBalance,
    required this.status,
  });

  factory CashOutResponseDto.fromJson(Map<String, dynamic> json) =>
      CashOutResponseDto(
        success: json['success'] as bool,
        cashOutId: json['cashOutId'] as String,
        coinAmount: (json['coinAmount'] as num).toInt(),
        cashValue: (json['cashValue'] as num).toDouble(),
        newBalance: (json['newBalance'] as num).toInt(),
        status: json['status'] as String,
      );
}

class CashOutHistoryDto {
  final String id;
  final int coinAmount;
  final double cashValue;
  final String paymentMethod;
  final String status;
  final DateTime requestedAt;
  final DateTime? completedAt;

  const CashOutHistoryDto({
    required this.id,
    required this.coinAmount,
    required this.cashValue,
    required this.paymentMethod,
    required this.status,
    required this.requestedAt,
    this.completedAt,
  });

  factory CashOutHistoryDto.fromJson(Map<String, dynamic> json) =>
      CashOutHistoryDto(
        id: json['id'] as String,
        coinAmount: (json['coinAmount'] as num).toInt(),
        cashValue: (json['cashValue'] as num).toDouble(),
        paymentMethod: json['paymentMethod'] as String,
        status: json['status'] as String,
        requestedAt: DateTime.parse(json['requestedAt'] as String),
        completedAt: json['completedAt'] != null
            ? DateTime.parse(json['completedAt'] as String)
            : null,
      );
}

File: backend/src/cashout/dto/cashout.dto.ts
import { ApiProperty } from '@nestjs/swagger';
import { IsString, IsInt, Min } from 'class-validator';
import { PaginationQueryDto, PaginationMeta } from '../../common/dto/pagination.dto';

export class CashOutRequestDto {
  @ApiProperty({ minimum: 10000 })
  @IsInt()
  @Min(10000)
  coinAmount: number;

  @ApiProperty({ description: 'paypal | bank_transfer | crypto' })
  @IsString()
  paymentMethod: string;

  @ApiProperty({ description: 'Email, IBAN, or wallet address depending on method' })
  @IsString()
  paymentDetails: string;
}

export class CashOutResponseDto {
  @ApiProperty()
  success: boolean;

  @ApiProperty()
  cashOutId: string;

  @ApiProperty()
  coinAmount: number;

  @ApiProperty()
  cashValue: number;

  @ApiProperty()
  newBalance: number;

  @ApiProperty({ description: 'pending | processing | completed | failed' })
  status: string;
}

export class CashOutHistoryDto {
  @ApiProperty()
  id: string;

  @ApiProperty()
  coinAmount: number;

  @ApiProperty()
  cashValue: number;

  @ApiProperty()
  paymentMethod: string;

  @ApiProperty({ description: 'pending | processing | completed | failed' })
  status: string;

  @ApiProperty()
  requestedAt: Date;

  @ApiProperty({ nullable: true })
  completedAt: Date | null;
}

export class CashOutHistoryQueryDto extends PaginationQueryDto {}

export class PaginatedCashOutHistoryDto {
  @ApiProperty({ type: [CashOutHistoryDto] })
  data: CashOutHistoryDto[];

  @ApiProperty()
  meta: PaginationMeta;
}
